# Merchants exploration

In the present notebook the whole set of initial categories in merchants.csv is analyzed. Other datasets such as historical_transactions.csv are imported to carry out exploration tasks on possible grouping columns. The present notebook does not perform any modification on the datasets and it is to be intended purely as a tool for data exploration.

## Import libraries and datasets

In [1]:
import os
import pandas as pd
import numpy as np
from collections import Counter

# Plot libraries
import matplotlib.pyplot as plt
import seaborn as sns
color = sns.color_palette()

# Random seed for reproducibility
seed = 202
np.random.seed(seed)

# Ignore warnings
import warnings
warnings.simplefilter('ignore')

# Garbage collector
import gc
gc.enable()

In [2]:
# Import datasets and check dimensions (historical_df will be used laterz)
merchants_df = pd.read_csv("merchants.csv")
historical_df = pd.read_csv("historical_transactions.csv")
print("merchants.csv shape : {}".format(merchants_df.shape))
print("historical_transaction.csv shape : {}".format(historical_df.shape))

merchants.csv shape : (334696, 22)
historical_transaction.csv shape : (29112361, 14)


In [3]:
merchants_df.head()

merchant_id  merchant_group_id  merchant_category_id  subsector_id  \
0  M_ID_838061e48c               8353                   792             9   
1  M_ID_9339d880ad               3184                   840            20   
2  M_ID_e726bbae1e                447                   690             1   
3  M_ID_a70e9c5f81               5026                   792             9   
4  M_ID_64456c37ce               2228                   222            21   

   numerical_1  numerical_2 category_1 most_recent_sales_range  \
0    -0.057471    -0.057471          N                       E   
1    -0.057471    -0.057471          N                       E   
2    -0.057471    -0.057471          N                       E   
3    -0.057471    -0.057471          Y                       E   
4    -0.057471    -0.057471          Y                       E   

  most_recent_purchases_range  avg_sales_lag3     ...      avg_sales_lag6  \
0                           E           -0.40     ...               -2.25   
1                           E           -0.72     ...               -0.74   
2                           E          -82.13     ...              -82.13   
3                           E             NaN     ...                 NaN   
4                           E             NaN     ...                 NaN   

   avg_purchases_lag6  active_months_lag6  avg_sales_lag12  \
0           18.666667                   6            -2.32   
1            1.291667                   6            -0.57   
2          260.000000                   2           -82.13   
3            4.666667                   6              NaN   
4            0.361111                   6              NaN   

   avg_purchases_lag12  active_months_lag12  category_4  city_id state_id  \
0            13.916667                   12           N      242        9   
1             1.687500                   12           N       22       16   
2           260.000000                    2           N       -1        5   
3             3.833333                   12           Y       -1       -1   
4             0.347222                   12           Y       -1       -1   

   category_2  
0         1.0  
1         1.0  
2         5.0  
3         NaN  
4         NaN  

[5 rows x 22 columns]

In [4]:
merchants_df.isnull().sum()

merchant_id                        0
merchant_group_id                  0
merchant_category_id               0
subsector_id                       0
numerical_1                        0
numerical_2                        0
category_1                         0
most_recent_sales_range            0
most_recent_purchases_range        0
avg_sales_lag3                    13
avg_purchases_lag3                 0
active_months_lag3                 0
avg_sales_lag6                    13
avg_purchases_lag6                 0
active_months_lag6                 0
avg_sales_lag12                   13
avg_purchases_lag12                0
active_months_lag12                0
category_4                         0
city_id                            0
state_id                           0
category_2                     11887
dtype: int64

In [5]:
def cat_plot(df, col):
    cnt_srs = df[col].value_counts()
    cnt_srs = cnt_srs.sort_index()
    plt.figure(figsize=(14,6))
    sns.barplot(cnt_srs.index, cnt_srs.values, alpha=0.8, color='green')
    plt.xticks(rotation='vertical')
    plt.xlabel('First active month', fontsize=12)
    plt.ylabel('Number of cards', fontsize=12)
    plt.title("First active month count in train set")
    plt.show()

In [6]:
print("Number of unique merchant_category_id in merchants.df :",merchants_df.merchant_category_id.nunique(), " out of rows : ",merchants_df.shape[0])

Number of unique merchant_category_id in merchants.df : 324  out of rows :  334696


## Features analysis

In the following lines we take into account each feature once at a time for exploration.

#### merchant_id

In [7]:
# Some merchant_ids are repeated more than once
print("Number of unique merchant_id in transactions set :",merchants_df.merchant_id.nunique(), " out of rows : ",merchants_df.shape[0])

Number of unique merchant_id in transactions set : 334633  out of rows :  334696


In [8]:
# Get how many times each recurrent merchant_id is repeated
count_series = merchants_df['merchant_id'].value_counts()
count_non_1 = count_series[count_series > 1]
print(count_non_1)

M_ID_bd49e37dda    4
M_ID_30340088f2    4
M_ID_1802942aaf    4
M_ID_992a180b15    4
M_ID_ef233cff26    4
M_ID_d123532c72    4
M_ID_42697d5d44    4
M_ID_6464db3b45    4
M_ID_ebbdb42da6    4
M_ID_c2b9ac2ea4    4
M_ID_dbbf07ebf0    4
M_ID_c470fbcfb9    2
M_ID_00a6ca8a8a    2
M_ID_508ecbdcde    2
M_ID_654509ddbc    2
M_ID_a8767b29ef    2
M_ID_32d73b8ecf    2
M_ID_fea38c640b    2
M_ID_7251df88ad    2
M_ID_a55beb9d9a    2
M_ID_0c4018d3a0    2
M_ID_8951e22556    2
M_ID_b794b9d9e8    2
M_ID_6c144405a0    2
M_ID_ae9fe1605a    2
M_ID_26d4fadb60    2
M_ID_0039220eb3    2
M_ID_0b8ed0c2b0    2
M_ID_1ceca881f0    2
M_ID_0a00fa9e8a    2
M_ID_49b779bc8a    2
M_ID_6017075769    2
M_ID_7483dde87d    2
M_ID_49daf8a838    2
M_ID_28c2aace87    2
M_ID_c0b712e11a    2
M_ID_57e97f6546    2
M_ID_9b0ef314cf    2
M_ID_645a6af169    2
M_ID_d2b5d4418d    2
M_ID_07a5671b4c    2
Name: merchant_id, dtype: int64


In [ ]:
count_series = merchants_df['merchant_id'].value_counts()
count_non_1 = count_series[count_series > 1]
print(count_non_1)

In [21]:
count_non_1.values

array([4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2,
       2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2])

We inspect the lines of the dataset corresponding to duplicated entries of merchant_id

In [9]:
merchants_df.loc[merchants_df.merchant_id.isin(count_non_1.index)]

merchant_id  merchant_group_id  merchant_category_id  \
53      M_ID_c0b712e11a                 35                   823   
54      M_ID_c0b712e11a                 35                   823   
111     M_ID_0039220eb3                 35                   836   
112     M_ID_0039220eb3                 35                   836   
3392    M_ID_bd49e37dda                 35                   692   
3393    M_ID_bd49e37dda               4170                   692   
3394    M_ID_bd49e37dda                 35                   692   
3395    M_ID_bd49e37dda               4170                   692   
4181    M_ID_ef233cff26                 35                   560   
4182    M_ID_ef233cff26              28799                   560   
4183    M_ID_ef233cff26              28799                   560   
4184    M_ID_ef233cff26                 35                   560   
7584    M_ID_dbbf07ebf0              69462                   278   
7585    M_ID_dbbf07ebf0                 35                   278   
7586    M_ID_dbbf07ebf0              69462                   278   
7587    M_ID_dbbf07ebf0                 35                   278   
8172    M_ID_0c4018d3a0                 35                   879   
8173    M_ID_0c4018d3a0                 35                   879   
8417    M_ID_c470fbcfb9                 35                   574   
8418    M_ID_c470fbcfb9                 35                   574   
11363   M_ID_508ecbdcde                 35                   650   
11364   M_ID_508ecbdcde                 35                   650   
24886   M_ID_0b8ed0c2b0                 35                   554   
24887   M_ID_0b8ed0c2b0                 35                   554   
25980   M_ID_49b779bc8a                 35                   451   
25981   M_ID_49b779bc8a                 35                   451   
29298   M_ID_8951e22556                 35                   434   
29299   M_ID_8951e22556                 35                   434   
29464   M_ID_30340088f2               3318                   544   
29465   M_ID_30340088f2                 35                   544   
...                 ...                ...                   ...   
182473  M_ID_57e97f6546                 35                   166   
182474  M_ID_57e97f6546                 35                   166   
197477  M_ID_6c144405a0                 35                   554   
197478  M_ID_6c144405a0                 35                   554   
210653  M_ID_c2b9ac2ea4              69907                   554   
210654  M_ID_c2b9ac2ea4                 35                   554   
210655  M_ID_c2b9ac2ea4                 35                   554   
210656  M_ID_c2b9ac2ea4              69907                   554   
221180  M_ID_992a180b15                 35                   554   
221181  M_ID_992a180b15               8568                   554   
221182  M_ID_992a180b15               8568                   554   
221183  M_ID_992a180b15                 35                   554   
228274  M_ID_d123532c72                 35                   385   
228275  M_ID_d123532c72              49094                   385   
228276  M_ID_d123532c72                 35                   385   
228277  M_ID_d123532c72              49094                   385   
326908  M_ID_d2b5d4418d                 35                   511   
326909  M_ID_d2b5d4418d                 35                   511   
330957  M_ID_42697d5d44                123                   690   
330958  M_ID_42697d5d44                 35                   690   
330959  M_ID_42697d5d44                123                   690   
330960  M_ID_42697d5d44                 35                   690   
333903  M_ID_6464db3b45               3604                   210   
333904  M_ID_6464db3b45                 35                   210   
333905  M_ID_6464db3b45               3604                   210   
333906  M_ID_6464db3b45                 35                   210   
334070  M_ID_1802942aaf                 35                   302   


merchant_id  merchant_group_id  merchant_category_id  \
110969  M_ID_00a6ca8a8a                 35                   511   
110970  M_ID_00a6ca8a8a                 35                   511   

        subsector_id  numerical_1  numerical_2 category_1  \
110969             7    -0.047556    -0.057471          Y   
110970             7    -0.057471    -0.057471          Y   

       most_recent_sales_range most_recent_purchases_range  avg_sales_lag3  \
110969                       E                           E            6.93   
110970                       A                           A            0.93   

           ...      avg_sales_lag6  avg_purchases_lag6  active_months_lag6  \
110969     ...                8.42           15.855769                   6   
110970     ...                0.86            0.871214                   6   

        avg_sales_lag12  avg_purchases_lag12  active_months_lag12  category_4  \
110969             8.57            16.219780                    7           Y   
110970             0.76             0.752641                   12           Y   

        city_id state_id  category_2  
110969       -1       -1         NaN  
110970       -1       -1         NaN  

[2 rows x 22 columns]

At first, I thought a good way to choose one representant for each merchant_id would have been to take the sample for which active_months_lag12 (or active_months_lag6) was the biggest: this way we would exclude the old entries and only consider those that are the most representative for the given merchant_id at the present time. The problem is that some merchant_id (for example M_ID_bd49e37dda, index = 3392 -> 3395) have several entries with the same active_month_lag (all of them) and still different features in other positions, for example merchant_group_id. Regarding merchant_group_id, we can notice that most of them seem to have value 35. Maybe this is the default value for the given merchant_group_id?

#### merchant_group_id

In [40]:
#check how many 35 are present in merchant_group_id
len(merchants_df.merchant_group_id.loc[merchants_df.merchant_group_id != 35])

#check each possible value
print(merchants_df.merchant_group_id.value_counts()[0:19])

35      46026
434       482
419       365
713       333
3648      305
48        283
1145      276
771       257
325       248
321       235
898       215
661       214
1476      213
123       210
3290      204
782       202
3014      196
863       192
5275      192
Name: merchant_group_id, dtype: int64


Ok, there are a lot of 35s and not so many other values, so probably 35 is the default value (it may mean "generic shop" or something similar). Here we have a choice to make: either we increase variance by taking out duplicated entries with merchant_group_id == 35, at the cost of introducing some bias, or we do the opposite taking out entries different from 35. At this point, I would probably go for the second option, since it seems more safe, but let's first see if the other features give us some information to guide the choice.

In [37]:
#check how many samples with same merchant_id and different merchant_group_id are present
single_el = 0;
multiple_el = 0;
for element in count_non_1.index:
    print("merchant_id = {}, value_counts_sum = \n {}".format(element, merchants_df.loc[merchants_df.merchant_id == element]['merchant_group_id'].value_counts()))
    if len(merchants_df.loc[merchants_df.merchant_id == element]['merchant_group_id'].value_counts()) == 1:
        single_el += 1;
    else:
        multiple_el +=1;
print("The amount of duplicated IDs corresponding to only one merchant_group_id is: {}".format(single_el))
print("The amount of duplicated IDs corresponding to multiple merchant_group_id is: {}".format(multiple_el))

merchant_id = M_ID_bd49e37dda, value_counts_sum = 
 35      2
4170    2
Name: merchant_group_id, dtype: int64
merchant_id = M_ID_30340088f2, value_counts_sum = 
 35      2
3318    2
Name: merchant_group_id, dtype: int64
merchant_id = M_ID_1802942aaf, value_counts_sum = 
 35       2
72963    2
Name: merchant_group_id, dtype: int64
merchant_id = M_ID_992a180b15, value_counts_sum = 
 35      2
8568    2
Name: merchant_group_id, dtype: int64
merchant_id = M_ID_ef233cff26, value_counts_sum = 
 35       2
28799    2
Name: merchant_group_id, dtype: int64
merchant_id = M_ID_d123532c72, value_counts_sum = 
 35       2
49094    2
Name: merchant_group_id, dtype: int64
merchant_id = M_ID_42697d5d44, value_counts_sum = 
 123    2
35     2
Name: merchant_group_id, dtype: int64
merchant_id = M_ID_6464db3b45, value_counts_sum = 
 35      2
3604    2
Name: merchant_group_id, dtype: int64
merchant_id = M_ID_ebbdb42da6, value_counts_sum = 
 35      2
1014    2
Name: merchant_group_id, dtype: int64
mercha

The previous result makes clear the fact that value 35 is present in almost all the fields interested by the problem of duplicated merchant_id. To be discussed.

#### merchant_category_id

Differently from what happened before, merchant_category_id doesn't seem to introduce problems with duplicated merchant_id. Let's confirm this observation with the same kind of analysis we used before.

In [38]:
#check how many samples with same merchant_id and different merchant_category_id are present
single_el = 0;
multiple_el = 0;
for element in count_non_1.index:
    print("merchant_id = {}, value_counts_sum = \n {}".format(element, merchants_df.loc[merchants_df.merchant_id == element]['merchant_category_id'].value_counts()))
    if len(merchants_df.loc[merchants_df.merchant_id == element]['merchant_category_id'].value_counts()) == 1:
        single_el += 1;
    else:
        multiple_el +=1;
print("The amount of duplicated IDs corresponding to only one merchant_category_id is: {}".format(single_el))
print("The amount of duplicated IDs corresponding to multiple merchant_category_id is: {}".format(multiple_el))

merchant_id = M_ID_bd49e37dda, value_counts_sum = 
 692    4
Name: merchant_category_id, dtype: int64
merchant_id = M_ID_30340088f2, value_counts_sum = 
 544    4
Name: merchant_category_id, dtype: int64
merchant_id = M_ID_1802942aaf, value_counts_sum = 
 302    4
Name: merchant_category_id, dtype: int64
merchant_id = M_ID_992a180b15, value_counts_sum = 
 554    4
Name: merchant_category_id, dtype: int64
merchant_id = M_ID_ef233cff26, value_counts_sum = 
 560    4
Name: merchant_category_id, dtype: int64
merchant_id = M_ID_d123532c72, value_counts_sum = 
 385    4
Name: merchant_category_id, dtype: int64
merchant_id = M_ID_42697d5d44, value_counts_sum = 
 690    4
Name: merchant_category_id, dtype: int64
merchant_id = M_ID_6464db3b45, value_counts_sum = 
 210    4
Name: merchant_category_id, dtype: int64
merchant_id = M_ID_ebbdb42da6, value_counts_sum = 
 383    4
Name: merchant_category_id, dtype: int64
merchant_id = M_ID_c2b9ac2ea4, value_counts_sum = 
 554    4
Name: merchant_catego

This stuff is getting deep... There is only one outlier, the other guys behave as expected. I feel that motherfucka does not have much to live.

#### subsector_id

This guy is present also in historical_transactions, so it is particularly interesting for us

In [31]:
#check how many samples with same merchant_id and different merchant_category_id are present
single_el = 0;
multiple_el = 0;
for element in count_non_1.index:
    if len(merchants_df.loc[merchants_df.merchant_id == element]['subsector_id'].value_counts()) == 1:
        single_el += 1;
    else:
        multiple_el +=1;
print("The amount of duplicated IDs corresponding to only one subsector_id is: {}".format(single_el))
print("The amount of duplicated IDs corresponding to multiple subsector_id is: {}".format(multiple_el))

The amount of duplicated IDs corresponding to only one subsector_id is: 41
The amount of duplicated IDs corresponding to multiple subsector_id is: 0


Oh sia reso grazie all'Altissimo!!! Sono univoci!!!

### Intermezzo

Ok, c'è chiaramente qualcosa che non va. Adesso ci fermiamo e cerchiamo di capire se per caso il nome merchant_id sia una bella trollata. In particolare, è possibile che il merchant_id non sia poi così univoco come ci hanno voluto far pensare, ma se vada piuttosto considerato come una parte di un ID completato dai campi merchant_group_id e merchant_category_id.

In [33]:
#Check characteristics of rows in historical_df having the same merchant_id as the "duplicated" rows in merchants_df
for element in count_non_1.index:
    print("merchant_id = {}, value_counts_sum = \n {}".format(element, sum(historical_df.loc[historical_df.merchant_id == element]['merchant_category_id'].value_counts().values)))

merchant_id = M_ID_bd49e37dda, value_counts_sum = 
 14
merchant_id = M_ID_30340088f2, value_counts_sum = 
 108
merchant_id = M_ID_1802942aaf, value_counts_sum = 
 1998
merchant_id = M_ID_992a180b15, value_counts_sum = 
 536
merchant_id = M_ID_ef233cff26, value_counts_sum = 
 171
merchant_id = M_ID_d123532c72, value_counts_sum = 
 796
merchant_id = M_ID_42697d5d44, value_counts_sum = 
 211
merchant_id = M_ID_6464db3b45, value_counts_sum = 
 4910
merchant_id = M_ID_ebbdb42da6, value_counts_sum = 
 2628
merchant_id = M_ID_c2b9ac2ea4, value_counts_sum = 
 668
merchant_id = M_ID_dbbf07ebf0, value_counts_sum = 
 189
merchant_id = M_ID_c470fbcfb9, value_counts_sum = 
 1859
merchant_id = M_ID_00a6ca8a8a, value_counts_sum = 
 1115097
merchant_id = M_ID_508ecbdcde, value_counts_sum = 
 2935
merchant_id = M_ID_654509ddbc, value_counts_sum = 
 234
merchant_id = M_ID_a8767b29ef, value_counts_sum = 
 194
merchant_id = M_ID_32d73b8ecf, value_counts_sum = 
 3494
merchant_id = M_ID_fea38c640b, value_co

In [ ]:
# Check the motherfucka
merchants_df.loc[merchants_df['merchant_id'] == 'M_ID_00a6ca8a8a']

Conclusione: basandoci su queste combinazioni di features, non sembra possibile avere un'aggregazione sensata. La soluzione potrebbe essere quella di ripetere tutti i diversi casi osservati facendo una merge per ogni combinazione dei sample e concatenando i risultati. In caso contrario, dovremmo decidere un rappresentante per ogni sample (in questo caso, staremmo supponendo che entry duplicate compaiano a causa di qualche errore di inserimento): un possibile metodo per eseguire questa operazione è basarsi sulla month_lag, però non tutti i sample duplicati sembrano distinguibili sulla base di tali features.
TODO: ragionare ancora sul secondo approccio.

## Other problematic features

#### avg_sales_lag,	avg_purchases_lag,	active_months_lag

The set of columns {avg_sales_lag, avg_purchases_lag, active_months_lag} present some problems. In particular, the fact that the values range from -1 to Inf. Let's first confirm this observation.

In [47]:
#3 months lags
print("Min avg_sales_lag3 = {}, Max avg_sales_lag3 = {}".format(min(merchants_df['avg_sales_lag3']), max(merchants_df['avg_sales_lag3'])))
print("Min avg_purchases_lag3 = {}, Max avg_purchases_lag3 = {}".format(min(merchants_df['avg_purchases_lag3']), max(merchants_df['avg_purchases_lag3'])))
print("Min active_months_lag3 = {}, Max active_months_lag3 = {} \n".format(min(merchants_df['active_months_lag3']), max(merchants_df['active_months_lag3'])))

#6 months lags
print("Min avg_sales_lag6 = {}, Max avg_sales_lag6 = {}".format(min(merchants_df['avg_sales_lag6']), max(merchants_df['avg_sales_lag6'])))
print("Min avg_purchases_lag6 = {}, Max avg_purchases_lag6 = {}".format(min(merchants_df['avg_purchases_lag6']), max(merchants_df['avg_purchases_lag6'])))
print("Min active_months_lag6 = {}, Max active_months_lag6 = {} \n".format(min(merchants_df['active_months_lag6']), max(merchants_df['active_months_lag6'])))

#12 months lags
print("Min avg_sales_lag12 = {}, Max avg_sales_lag12 = {}".format(min(merchants_df['avg_sales_lag12']), max(merchants_df['avg_sales_lag12'])))
print("Min avg_purchases_lag12 = {}, Max avg_purchases_lag12 = {}".format(min(merchants_df['avg_purchases_lag12']), max(merchants_df['avg_purchases_lag12'])))
print("Min active_months_lag12 = {}, Max active_months_lag12 = {} \n".format(min(merchants_df['active_months_lag12']), max(merchants_df['active_months_lag12'])))

Min avg_sales_lag3 = -82.13, Max avg_sales_lag3 = 851844.64
Min avg_purchases_lag3 = 0.33349533, Max avg_purchases_lag3 = inf
Min active_months_lag3 = 1, Max active_months_lag3 = 3 

Min avg_sales_lag6 = -82.13, Max avg_sales_lag6 = 1513959.0
Min avg_purchases_lag6 = 0.16704466, Max avg_purchases_lag6 = inf
Min active_months_lag6 = 1, Max active_months_lag6 = 6 

Min avg_sales_lag12 = -82.13, Max avg_sales_lag12 = 2567408.0
Min avg_purchases_lag12 = 0.09832954, Max avg_purchases_lag12 = inf
Min active_months_lag12 = 1, Max active_months_lag12 = 12 



In [56]:
#check how many inf values are present in avg_purchases_lag
print("Number of inf values in avg_purchases_lag3: {} \n".format(len(merchants_df[merchants_df == np.inf]['avg_sales_lag3'])))
print("Number of inf values in avg_purchases_lag6: {} \n".format(len(merchants_df[merchants_df == np.inf]['avg_sales_lag6'])))
print("Number of inf values in avg_purchases_lag12: {} \n".format(len(merchants_df[merchants_df == np.inf]['avg_sales_lag12'])))

Number of inf values in avg_purchases_lag3: 334696 

Number of inf values in avg_purchases_lag6: 334696 

Number of inf values in avg_purchases_lag12: 334696 



In [51]:
#check max value in avg_purchase_lag where avg_purchase_lag != inf
print("Min avg_purchases_lag3 = {}, Max avg_purchases_lag3 = {}".format(min(merchants_df['avg_purchases_lag3']), max(merchants_df[merchants_df != np.inf]['avg_purchases_lag3'])))
print("Min avg_purchases_lag6 = {}, Max avg_purchases_lag6 = {}".format(min(merchants_df['avg_purchases_lag6']), max(merchants_df[merchants_df != np.inf]['avg_purchases_lag6'])))
print("Min avg_purchases_lag12 = {}, Max avg_purchases_lag12 = {}".format(min(merchants_df['avg_purchases_lag12']), max(merchants_df[merchants_df != np.inf]['avg_purchases_lag12'])))

Min avg_purchases_lag3 = 0.33349533, Max avg_purchases_lag3 = 61851.33333333
Min avg_purchases_lag6 = 0.16704466, Max avg_purchases_lag6 = 56077.5
Min avg_purchases_lag12 = 0.09832954, Max avg_purchases_lag12 = 50215.55555556


In [42]:
#substitute inf values with 2*max_value_in_column
for i in ['3', '6', '12']:
    max_val = max(merchants_df[merchants_df != np.inf]['avg_purchases_lag' + i])
    df.replace([np.inf, -np.inf], 2*max_val)

merchant_id                        0
merchant_group_id                  0
merchant_category_id               0
subsector_id                       0
numerical_1                        0
numerical_2                        0
category_1                         0
most_recent_sales_range            0
most_recent_purchases_range        0
avg_sales_lag3                    13
avg_purchases_lag3                 0
active_months_lag3                 0
avg_sales_lag6                    13
avg_purchases_lag6                 0
active_months_lag6                 0
avg_sales_lag12                   13
avg_purchases_lag12                0
active_months_lag12                0
category_4                         0
city_id                            0
state_id                           0
category_2                     11887
dtype: int64

### Export cleaned dataframe

In [60]:
merchants_df.to_csv('(1)merchants.csv', index = False)